test/test_common.c

```diff
@@ -412,6 +436,34 @@ int getRedisPort(void) {
     assert_true(currRedisInst>=0);
     return redisPort[currRedisInst];
 }
+
+void setValgrind() {
+    useValgrind = 1;
+}
+
+void checkValgrindLog(const char *filename) {
+    const char *expectedSummary = "ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)\n";
```

Comment:
i'm not certain this is enough (redis uses another approach, see `find_valgrind_errors` which isn't perfect either)
please try to create several types of errors to validates it catches them.
* leak
* invalid read / write (like access to a buffer outside it's range)
* uninitialized value read (maybe a conditional jump)
* using memcpy to move data in a buffer instead of memmove

Reply:
also, this can break when a suppression is added, and then we'll need to change it (and hopefully remember to re-test thees), so maybe add a suppression mechanism now, and validate the above afterwards.

Reply:
I think it is sufficient. The summary truly reflects what is considered for an error based on what is configured for valgrind vs what is not (suppresssed).

**Example to Leak report:**
```
==160590== Memcheck, a memory error detector
==160590== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==160590== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==160590== Command: ./my_program
==160590== 
==160590== 
==160590== HEAP SUMMARY:
==160590==     in use at exit: 20 bytes in 1 blocks
==160590==   total heap usage: 1 allocs, 0 frees, 20 bytes allocated
==160590== 
==160590== 20 bytes in 1 blocks are definitely lost in loss record 1 of 1
==160590==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==160590==    by 0x10915E: main (my_program.c:4)
==160590== 
==160590== LEAK SUMMARY:
==160590==    definitely lost: 20 bytes in 1 blocks
==160590==    indirectly lost: 0 bytes in 0 blocks
==160590==      possibly lost: 0 bytes in 0 blocks
==160590==    still reachable: 0 bytes in 0 blocks
==160590==         suppressed: 0 bytes in 0 blocks
==160590== 
==160590== For lists of detected and suppressed errors, rerun with: -s
==160590== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```
**Example to invalid read:**
```
==160657== Memcheck, a memory error detector
==160657== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==160657== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==160657== Command: ./invalid_read
==160657== 
==160657== Invalid read of size 4
==160657==    at 0x1091D3: main (invalid_read.c:13)
==160657==  Address 0x4a59054 is 0 bytes after a block of size 20 alloc'd
==160657==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==160657==    by 0x10919E: main (invalid_read.c:5)
==160657== 
Value: 0
==160657== 
==160657== HEAP SUMMARY:
==160657==     in use at exit: 0 bytes in 0 blocks
==160657==   total heap usage: 2 allocs, 2 frees, 1,044 bytes allocated
==160657== 
==160657== All heap blocks were freed -- no leaks are possible
==160657== 
==160657== For lists of detected and suppressed errors, rerun with: -s
==160657== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

**Example to uninitialized value read:**
```
==160716== Memcheck, a memory error detector
==160716== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==160716== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==160716== Command: ./uninitialized
==160716== 
==160716== Conditional jump or move depends on uninitialised value(s)
==160716==    at 0x48DB958: __vfprintf_internal (vfprintf-internal.c:1687)
==160716==    by 0x48C5D3E: printf (printf.c:33)
==160716==    by 0x10916A: main (in /tmp/valgrind/uninit/uninitialized)
==160716== 
==160716== Use of uninitialised value of size 8
==160716==    at 0x48BF69B: _itoa_word (_itoa.c:179)
==160716==    by 0x48DB574: __vfprintf_internal (vfprintf-internal.c:1687)
==160716==    by 0x48C5D3E: printf (printf.c:33)
==160716==    by 0x10916A: main (in /tmp/valgrind/uninit/uninitialized)
==160716== 
==160716== Conditional jump or move depends on uninitialised value(s)
==160716==    at 0x48BF6AD: _itoa_word (_itoa.c:179)
==160716==    by 0x48DB574: __vfprintf_internal (vfprintf-internal.c:1687)
==160716==    by 0x48C5D3E: printf (printf.c:33)
==160716==    by 0x10916A: main (in /tmp/valgrind/uninit/uninitialized)
==160716== 
==160716== Conditional jump or move depends on uninitialised value(s)
==160716==    at 0x48DC228: __vfprintf_internal (vfprintf-internal.c:1687)
==160716==    by 0x48C5D3E: printf (printf.c:33)
==160716==    by 0x10916A: main (in /tmp/valgrind/uninit/uninitialized)
==160716== 
==160716== Conditional jump or move depends on uninitialised value(s)
==160716==    at 0x48DB6EE: __vfprintf_internal (vfprintf-internal.c:1687)
==160716==    by 0x48C5D3E: printf (printf.c:33)
==160716==    by 0x10916A: main (in /tmp/valgrind/uninit/uninitialized)
==160716== 
Uninitialized value: 0
==160716== 
==160716== HEAP SUMMARY:
==160716==     in use at exit: 0 bytes in 0 blocks
==160716==   total heap usage: 1 allocs, 1 frees, 1,024 bytes allocated
==160716== 
==160716== All heap blocks were freed -- no leaks are possible
==160716== 
==160716== Use --track-origins=yes to see where uninitialised values come from
==160716== For lists of detected and suppressed errors, rerun with: -s
==160716== ERROR SUMMARY: 5 errors from 5 contexts (suppressed: 0 from 0)
```
**Example to using memcpy to move data in a buffer instead of memmove**
Failed to reproduce. Didn't make much attempts, assuming it's the same.

Reply:
Note that right now it is more strict regarding suppressed, but we can relaxi it by need. i.e.:
"ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)\n";
-->
"ERROR SUMMARY: 0 errors from 0 contexts ";

Reply:
Ok. Regarding the suppressing I'm not sure how it'll be represented. Please check.

Reply:
```
==327641== Memcheck, a memory error detector
==327641== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==327641== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==327641== Command: ./my_program
==327641== 
==327641== 
==327641== HEAP SUMMARY:
==327641==     in use at exit: 20 bytes in 1 blocks
==327641==   total heap usage: 1 allocs, 0 frees, 20 bytes allocated
==327641== 
==327641== LEAK SUMMARY:
==327641==    definitely lost: 0 bytes in 0 blocks
==327641==    indirectly lost: 0 bytes in 0 blocks
==327641==      possibly lost: 0 bytes in 0 blocks
==327641==    still reachable: 0 bytes in 0 blocks
==327641==         suppressed: 20 bytes in 1 blocks
==327641== 
==327641== For lists of detected and suppressed errors, rerun with: -s
==327641== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 1 from 1)
```

It is rather clear what it takes to modify in case tomorrow errors will be suppressed. 
But as long as we have none, it is better to stay strict and verify that none was 
suppressed as well.